### DISCUSSION TITLE
Failed to import: /opt/airflow/dags/taxi_pipeline_solution.py

### DISCUSSION THREAD

**Discussion by alexperry-shifu** [apache/airflow#19122] [Q&A]
### Apache Airflow version

2.2.0 (latest released)

### Operating System

apache/airflow:2.2.0-python3.6

### Deployment

Other Docker-based deployment

### Deployment details

Docker Compose & Docker Swarm

### What happened

My software versions:

Apache Airflow version: 2.2.0
TensorFlow version: 2.6.0
TFX Version: 1.3.0
Python version: 3.6

Hi,

My apologies if I have submitted this to the wrong site.

I want to run the taxi pipeline solution example dag on a distributed Apache Airflow cluster. I successfully built an Apache Airflow cluster on physical hosts using the docker image apache/airflow:2.2.0-python3.6 with Docker Swarm. The example dags in Apache Airflow seem to work fine. I made only the following changes to taxi_pipeline_solution.py to set the root path correctly:

_taxi_root = '/opt/airflow' #os.path.join(os.environ['HOME'], 'airflow')

I tried run the taxi pipeline solution example dag via docker-compose and it worked about 3-of-4 runs.
I tried run the taxi pipeline solution example dag via Docker Swarm on a single host and it worked about 3-of-4 runs.
I tried run the taxi pipeline solution example dag via Docker Swarm on a multiples hosts with a single worker and the dag worked 0-of-4 runs. The failures occurred randomly on different components (i.e. CsvExampleGen, Transform, latest_blessed_model_resolver, etc...) but with the same error:

**tensorflow.python.framework.errors_impl.NotFoundError: /opt/airflow/tfx/pipelines/taxi_solution/_wheels/tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a-py3-none-any.whl.tmp; No such file or directory**

Has anyone seen this before? Does anyone have an example TFX dag that works on an Apache Airflow cluster? The random nature of which component fails makes me wonder if there is a timing issue I am missing.

Any thoughts, tips, our guidance would be greatly appreciated.

The full stack trace is below:

tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | Copying tfx_user_code_Transform.egg-info to /tmp/tmp5hfsb7xp/tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a-py3.6.egg-info
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | Copying tfx_user_code_Transform.egg-info to /tmp/tmppe0rm7e1/tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a-py3.6.egg-info
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | running install_scripts
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | running install_scripts
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | creating /tmp/tmp5hfsb7xp/tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.dist-info/WHEEL
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | creating /tmp/tmppe0rm7e1/tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.dist-info/WHEEL
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | creating '/tmp/tmp4d285bx_/tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a-py3-none-any.whl' and adding '/tmp/tmp5hfsb7xp' to it
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | creating '/tmp/tmppuomc7al/tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a-py3-none-any.whl' and adding '/tmp/tmppe0rm7e1' to it
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'taxi_pipeline.py'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'taxi_pipeline.py'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'taxi_pipeline_solution.py'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'taxi_pipeline_solution.py'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'taxi_utils.py'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'taxi_utils.py'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'taxi_utils_solution.py'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'taxi_utils_solution.py'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.dist-info/METADATA'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.dist-info/METADATA'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.dist-info/WHEEL'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.dist-info/WHEEL'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.dist-info/top_level.txt'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.dist-info/top_level.txt'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.dist-info/RECORD'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | adding 'tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.dist-info/RECORD'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | removing /tmp/tmp5hfsb7xp
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | removing /tmp/tmppe0rm7e1
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | [2021-10-18 17:56:31,764: INFO/ForkPoolWorker-15] Successfully built user code wheel distribution at '/opt/airflow/tfx/pipelines/taxi_solution/_wheels/tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a-py3-none-any.whl'; target user module is 'taxi_utils_solution'.
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | [2021-10-18 17:56:31,765: INFO/ForkPoolWorker-15] Full user module path is 'taxi_utils_solution@/opt/airflow/tfx/pipelines/taxi_solution/_wheels/tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a-py3-none-any.whl'
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | [2021-10-18 17:56:31,765: ERROR/ForkPoolWorker-16] Failed to import: /opt/airflow/dags/taxi_pipeline_solution.py
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | Traceback (most recent call last):
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/tfx/dsl/io/plugins/tensorflow_gfile.py", line 90, in rename
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | tf.io.gfile.rename(src, dst, overwrite=overwrite)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/tensorflow/python/lib/io/file_io.py", line 624, in rename_v2
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | compat.path_to_bytes(src), compat.path_to_bytes(dst), overwrite)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | tensorflow.python.framework.errors_impl.NotFoundError: /opt/airflow/tfx/pipelines/taxi_solution/_wheels/tfx_user_code_Transform-0.0+9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a-py3-none-any.whl.tmp; No such file or directory
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname |
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | The above exception was the direct cause of the following exception:
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname |
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | Traceback (most recent call last):
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/airflow/models/dagbag.py", line 331, in _load_modules_from_file
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | loader.exec_module(new_module)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "", line 678, in exec_module
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "", line 219, in _call_with_frames_removed
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/opt/airflow/dags/taxi_pipeline_solution.py", line 188, in
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | beam_pipeline_args=_beam_pipeline_args))
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/tfx/orchestration/airflow/airflow_dag_runner.py", line 96, in run
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | tfx_pipeline.pipeline_info.pipeline_root)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/tfx/dsl/components/base/base_component.py", line 193, in _resolve_pip_dependencies
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | resolved_dependency = dependency.resolve(pipeline_root)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/tfx/components/util/udf_utils.py", line 150, in resolve
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | self.component.id, module_file, pipeline_root)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/tfx/components/util/udf_utils.py", line 269, in package_user_module_file
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | fileio.rename(temp_dist_file_path, dist_file_path, overwrite=True)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/tfx/dsl/io/fileio.py", line 96, in rename
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | src_fs.rename(src, dst, overwrite=overwrite)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/tfx/dsl/io/plugins/tensorflow_gfile.py", line 92, in rename
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | raise filesystem.NotFoundError() from e
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | tfx.dsl.io.filesystem.NotFoundError
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | [2021-10-18 17:56:31,769: INFO/ForkPoolWorker-15] Generating ephemeral wheel package for '/opt/airflow/dags/taxi_utils_solution.py' (including modules: ['taxi_pipeline_solution', 'taxi_utils', 'taxi_utils_solution', 'taxi_pipeline']).
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | [2021-10-18 17:56:31,770: INFO/ForkPoolWorker-15] User module package has hash fingerprint version 9e51c5beb5701e5d61fbb47361d9b14d938da17b36161d417af64125ba6eb02a.
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | [2021-10-18 17:56:31,771: INFO/ForkPoolWorker-15] Executing: ['/usr/local/bin/python', '/tmp/tmppeaq9sv3/_tfx_generated_setup.py', 'bdist_wheel', '--bdist-dir', '/tmp/tmp72g30zpg', '--dist-dir', '/tmp/tmp49kzg946']
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | [2021-10-18 17:56:31,917: ERROR/ForkPoolWorker-16] Failed to execute task dag_id could not be found: taxi_solution. Either the dag did not exist or it failed to parse..
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | Traceback (most recent call last):
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/airflow/executors/celery_executor.py", line 121, in _execute_in_fork
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | args.func(args)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/airflow/cli/cli_parser.py", line 48, in command
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | return func(*args, **kwargs)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/airflow/utils/cli.py", line 92, in wrapper
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | return f(*args, **kwargs)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/airflow/cli/commands/task_command.py", line 276, in task_run
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | dag = get_dag(args.subdir, args.dag_id)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/airflow/utils/cli.py", line 195, in get_dag
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | 'parse.'.format(dag_id)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | airflow.exceptions.AirflowException: dag_id could not be found: taxi_solution. Either the dag did not exist or it failed to parse.
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | [2021-10-18 17:56:32,037: ERROR/ForkPoolWorker-16] Task airflow.executors.celery_executor.execute_command[35fcdfa9-b730-47ff-9f4c-e9301ccafe35] raised unexpected: AirflowException('Celery command failed on host: tfx-airflow-tutorial-worker-1',)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | Traceback (most recent call last):
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/celery/app/trace.py", line 450, in trace_task
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | R = retval = fun(*args, **kwargs)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/celery/app/trace.py", line 731, in protected_call
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | return self.run(*args, **kwargs)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/airflow/executors/celery_executor.py", line 90, in execute_command
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | _execute_in_fork(command_to_exec, celery_task_id)
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | File "/home/airflow/.local/lib/python3.6/site-packages/airflow/executors/celery_executor.py", line 101, in _execute_in_fork
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | raise AirflowException('Celery command failed on host: ' + get_hostname())
tfx-airflow-tutorial-cluster-worker-1.1.rj2rl9f3jn74@hostname | airflow.exceptions.AirflowException: Celery command failed on host: tfx-airflow-tutorial-worker-1

### What you expected to happen

I expected the taxi_pipeline_solution.py dag to run successfully.

I suspect I have stumbled on a race condition or there is a timing parameter I am not setting.

### How to reproduce

Attempt to run the standard TensorFlow Extended example w/ the Apache Air using the versions I have specified.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by uranusjr** [comment]
This is a DAG file written by you, not Airflow, so it's not a bug.
